src/library_browser.js

```diff
@@ -766,57 +766,75 @@ var LibraryBrowser = {
+  emscripten_async_load_script_promise: (url) => {
     url = UTF8ToString(url);
-    onload = {{{ makeDynCall('v', 'onload') }}};
-    onerror = {{{ makeDynCall('v', 'onerror') }}};
 
-#if PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) {
-      err(`emscripten_async_load_script("${url}") failed, emscripten_async_load_script is currently not available in pthreads!`);
-      return onerror ? onerror() : undefined;
-    }
-#endif
-#if ASSERTIONS
-    assert(runDependencies === 0, 'async_load_script must be run when no other dependencies are active');
-#endif
-    {{{ runtimeKeepalivePush() }}}
+    // Creating a promise for script loading functionality
+    return new Promise((resolve, reject) => {
+    
+      #if PTHREADS
+      if (ENVIRONMENT_IS_PTHREAD) {
+        err(`emscripten_async_load_script("${url}") failed, emscripten_async_load_script is currently not available in pthreads!`);
+        return reject('emscripten_async_load_script is not available for pthreads');
+      }
+      #endif
+      
+      #if ASSERTIONS
+      assert(runDependencies === 0, 'async_load_script must be run when no other dependencies are active');
+      #endif
+      
+      {{{ runtimeKeepalivePush() }}}
 
-    var loadDone = () => {
-      {{{ runtimeKeepalivePop() }}}
-      if (onload) {
-        if (runDependencies > 0) {
-          dependenciesFulfilled = onload;
-        } else {
-          onload();
-        }
+      var loadDone = () => {
+        {{{ runtimeKeepalivePop() }}}
+        resolve();
       }
-    }
 
-    var loadError = () => {
-      {{{ runtimeKeepalivePop() }}}
-      onerror?.();
-    };
+      var loadError = () => {
+        {{{ runtimeKeepalivePop() }}}
+        reject();
+      };
 
-#if ENVIRONMENT_MAY_BE_NODE && DYNAMIC_EXECUTION
-    if (ENVIRONMENT_IS_NODE) {
-      readAsync(url, (data) => {
-        eval(data);
-        loadDone();
-      }, loadError, false);
-      return;
-    }
-#endif
+      #if ENVIRONMENT_MAY_BE_NODE && DYNAMIC_EXECUTION
+      if (ENVIRONMENT_IS_NODE) {
+        readAsync(url, (data) => {
+          eval(data);
+          loadDone();
+        }, loadError, false);
+        return;
+      }
+      #endif
+
+      var script = document.createElement('script');
+      script.onload = loadDone;
+      script.onerror = loadError;
+      script.src = url;
+      document.body.appendChild(script);
 
-    var script = document.createElement('script');
-    script.onload = loadDone;
-    script.onerror = loadError;
-    script.src = url;
-    document.body.appendChild(script);
+    });
   },
 
+
+  // Refactored the function to implement the promise based functionality and backward compatibility with callbacks
+  emscripten_async_load_script__deps: ['emscripten_async_load_script_promise', '$UTF8ToString'],
+  emscripten_async_load_script: function(url, onload, onerror)  {
+      
+      emscripten_async_load_script_promise(url).then(() => {
+        if (onload){
+          onload();
```

Comment:
You can't call these `onload` and `onerro` callback backs directly link this I think.  These arguments are C pointers (numbers) on not JS function.   I think you need to do what the old code did with `{{{ makeDynCall('v', 'onload') }}};`.

Reply:
@sbc100  converted calls into the old form